### PR TITLE
fix mypy typing errors in pytorch_lightning/strategies/parallel.py

### PR DESCRIPTION
**Pull request by Cyprien-Ricque**
## What does this PR do?

Fixes mypy typing errors in pytorch_lightning/strategies/parallel.py in https://github.com/Lightning-AI/lightning/issues/13445.

- Adding a new protocol ```DistributedDataParallel``` to inform mypy of the existence of the function ```no_sync``` in ```torch.nn.parallel.DistributedDataParallel```
- Usage of ``assert``` to check non-null and types.

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by Borda — APPROVED** [review]

File: src/pytorch_lightning/strategies/parallel.py [hunk 1]
```diff
@@ -81,16 +80,18 @@ def is_global_zero(self) -> bool:
...
 
     @parallel_devices.setter
-    def parallel_devices(self, parallel_devices):
+    def parallel_devices(self, parallel_devices: Optional[List[torch.device]]) -> None:
         self._parallel_devices = parallel_devices
 
     @property
-    def distributed_sampler_kwargs(self):
-        distributed_sampler_kwargs = dict(num_replicas=len(self.parallel_devices), rank=self.global_rank)
+    def distributed_sampler_kwargs(self) -> Dict[str, int]:
```
Comment (awaelchli) [inline src/pytorch_lightning/strategies/parallel.py:91]:
```suggestion
    def distributed_sampler_kwargs(self) -> Dict[str, Any]:
```
This would have to be more general I think, because subclasses could set different kwargs, in general?

**Review by justusschock — APPROVED** [review]

**Review by otaj — APPROVED** [review]

File: src/pytorch_lightning/strategies/parallel.py [hunk 2]
```diff
@@ -118,18 +119,21 @@ def reduce_boolean_decision(self, decision: bool) -> bool:
...
+    def block_backward_sync(self) -> Generator:
         """Blocks ddp sync gradients behaviour on backwards pass.
 
         This is useful for skipping sync when accumulating gradients, reducing communication overhead
         Returns: context manager with sync behaviour off
         """
-        if isinstance(self.model, DistributedDataParallel):
+        if isinstance(self.model, torch.nn.parallel.DistributedDataParallel) and isinstance(
+            self.model, pl.utilities.types.DistributedDataParallel
+        ):
```
Comment (awaelchli) [inline src/pytorch_lightning/strategies/parallel.py:130]:
```suggestion
        if isinstance(self.model, pl.utilities.types.DistributedDataParallel):
```

**Review by rohitgr7 — APPROVED** [review]

**Review by carmocca — APPROVED** [review]
